### PR TITLE
Validate AppDomainSetup.ApplicationBase path more like .NET does

### PR DESCRIPTION
**Pull request by ntherning**
On Windows we should check the path for invalid characters. Also expand relative paths set using file:// URLs like .NET does. This patch also removes the code which prepends "//" on Windows when a file:// URL is given as .NET doesn't seem to do that in more recent versions.

### CODE REVIEW

**Comment by ntherning** [comment]
Opted to not attempt to use the `AppDomainSetup` class from the referencesource as it would be a lot of work for something which is a corner case (`file://` URLs). I fixed up the test a bit to be more explicit about what is the expected behaviour on Windows vs other platforms.

**Comment by ntherning** [comment]
I have run the modified AppDomainSetupTest on both Windows (MSVC) and OS X and all is green.

File: mcs/class/corlib/System/AppDomainSetup.cs [hunk 1]
```diff
@@ -139,14 +140,18 @@ static string GetAppBase (string appBase)
...
+			if (Path.DirectorySeparatorChar != '/') {
+				if (appBase.IndexOf (':', appBase.StartsWith (@"\\?\") ? 6 : 2) != -1) {
+					throw new NotSupportedException ("The given path's format is not supported.");
 				}
-			} else {
-				appBase = Path.GetFullPath (appBase);
 			}
 
+			// Trigger a validation of the path
+			new FileIOPermission (FileIOPermissionAccess.PathDiscovery, appBase).Demand ();
```
Comment (marek-safar) [inline mcs/class/corlib/System/AppDomainSetup.cs]: Don't bother with CAS

File: mcs/class/corlib/System/AppDomainSetup.cs [hunk 2]
```diff
@@ -139,14 +140,18 @@ static string GetAppBase (string appBase)
...
 				if (Path.DirectorySeparatorChar != '/')
 					appBase = appBase.Replace ('/', Path.DirectorySeparatorChar);
-				if (Environment.IsRunningOnWindows) {
-					// Under Windows prepend "//" to indicate it's a local file
-					appBase = "//" + appBase;
+			}
+			appBase = Path.GetFullPath (appBase);
+
+			if (Path.DirectorySeparatorChar != '/') {
+				if (appBase.IndexOf (':', appBase.StartsWith (@"\\?\") ? 6 : 2) != -1) {
```
Comment (marek-safar) [inline mcs/class/corlib/System/AppDomainSetup.cs]: StartsWith needs StringComparison.Ordinal

**Comment by ntherning** [comment]
Thanks for taking a look! I have updated the PR as per your comments.

File: mcs/class/corlib/System/AppDomainSetup.cs [hunk 3]
```diff
@@ -139,14 +140,19 @@ static string GetAppBase (string appBase)
...
+				bool isExtendedPath = appBase.StartsWith (@"\\?\", StringComparison.Ordinal);
+				if (appBase.IndexOf (':', isExtendedPath ? 6 : 2) != -1) {
+					throw new NotSupportedException ("The given path's format is not supported.");
 				}
-			} else {
-				appBase = Path.GetFullPath (appBase);
 			}
 
+			// Trigger a validation of the path
+			new FileIOPermission (FileIOPermissionAccess.PathDiscovery, appBase);
```
Comment (marek-safar) [inline mcs/class/corlib/System/AppDomainSetup.cs]: Better not to add CAS

Comment (ntherning) [inline mcs/class/corlib/System/AppDomainSetup.cs]: Please clarify, I'm not sure what CAS means in this context. The call to the FileIOPermission() constructor triggers a validation of the path, checking for invalid path characters such as ? etc on Windows. If there's a better way to do this using other APIs then please let me know.

Comment (marek-safar) [inline mcs/class/corlib/System/AppDomainSetup.cs]: CodeAccessSecurity is available on few profiles only. You need to guard it with FEATURE_MONO_CAS

Comment (ntherning) [inline mcs/class/corlib/System/AppDomainSetup.cs]: I see. You might have missed it but I removed the call to `Demand()` so AFAICT no security check is performed. The only thing I'm interested in is invoking the FileIOPermission() constructor which validates the path characters. Are you saying that the FileIOPermission class is unavailable unless FEATURE_MONO_CAS is defined? On Windows FEATURE_MONO_CAS doesn't seem to be defined but invoking the constructor seems to work just fine.

Comment (marek-safar) [inline mcs/class/corlib/System/AppDomainSetup.cs]: It's not available on linked profiles although that should be fine as they don't usually support appdomains either

Comment (ntherning) [inline mcs/class/corlib/System/AppDomainSetup.cs]: On Windows FEATURE_MONO_CAS doesn't seem to be defined. If I add 
```
#if FEATURE_MONO_CAS
   ...
#endif
```

around the `FileIOPermission()` call it doesn't seem to be included and the test fails. How do I make sure it's enabled on Windows?
